### DISCUSSION TITLE
Enable 2 way ssl with emqx

### DISCUSSION THREAD

**Discussion by sushantAtAws** [emqx/emqx#5233] [Q&A] [answered]
Hi, 
I am following this tutorial to enable 2 way ssl btw the broker and client: https://www.emqx.io/blog/enable-two-way-ssl-for-emqx

I uncommented this config in the conf file and also provided the server key, cert and ca. Also restarted the broker.
`listener.ssl.external.verify = verify_peer`

I notice that a client can still connect and publish over the port 8883 without providing the client certificate/key.

Here is the command I used:
`mosquitto_pub -i bla2 -t test -m "{\"user\":$USER}" -d  -p 8883 --cafile ~/certs/ca.pem -h 127.0.0.1`
Result:
```
Client bla2 sending CONNECT
Client bla2 received CONNACK (0)
Client bla2 sending PUBLISH (d0, q0, r0, m1, 'test', ... (15 bytes))
Client bla2 sending DISCONNECT
```

If I don't provide the ca pem, it fails to connect:
```
mosquitto_pub -i bla2 -t test -m "{\"user\":$USER}" -d  -p 8883
Client bla2 sending CONNECT
Error: The connection was lost.
```
This tells me that the one way ssl is working (client validating the server cert) but not the other way.

Am I missing something?

**Comment by sushantAtAws** [comment, marked as answer]
Nvm, I was missing the `fail_if_no_peer_cert` option to make it fail when the certs are missing.

**Reply by JeffreyUrban** [reply]
The blog post appears to not establish knowledge of the client cert on the server. See [this issue](https://github.com/emqx/blog/issues/142). Therefore, it seems that the `fail_if_no_peer_cert` option would require only that the client have _any_ cert, and not that it had a known and allowed cert.

**Comment by Suryaobulareddy** [comment]
Hi @sushantAtAws,

I'm trying to setup the same thing, what you tried earlier but no luck

If you help me It would be great thing

Thank You

https://github.com/emqx/emqx/discussions/6213

**Reply by zmstone** [reply]
For the record: This has been answered
